### PR TITLE
Implement convergence type recipe for FittedPumasModel

### PR DESCRIPTION
**Pull request by asinghvi17**
Implements a Plots type recipe, which plots the convergence.  This can be changed to a user recipe `convergence` or something like that.

### CODE REVIEW

**Comment by asinghvi17** [comment]
I've changed the plotting API to reflect that plotting in Pumas will need to be extensible.  Therefore, my proposal for the internal API is this:
- Each "type" will implement methods to access plottable data which are required for the various plot types.  This will allow the creation of an extensible plotting framework, which can be used by any type which has the required methods implemented for it.
- These functions will be documented for each plot type, so that using them later becomes easy. 
- Each plot type (convergence, qq, etc. ) will then be able to use "multiple dispatch" to get plottable data easily.  This will simplify recipes and decrease the number of changes that need to be made.

We also need to look at how we can avoid conflicts with StatsPlots if possible (no overlapping recipes, et cetera).

**Comment by asinghvi17** [comment]
Apologies for the dual commits, should this be squashed before merging?

**Comment by pkofod** [comment]
> Apologies for the dual commits, should this be squashed before merging?

Don't worry about that.

File: src/estimation/diagnostics.jl [hunk 1]
```diff
@@ -481,3 +481,106 @@ function DataFrames.DataFrame(i::FittedPumasModelInspection; include_covariates=
...
+################################################################################
+#                              Plotting functions                              #
+################################################################################
+
+########################################
+#   Convergence plot infrastructure    #
+########################################
+
+"""
+    _observedvalues(obj)
```
Comment (vjd) [inline src/estimation/diagnostics.jl]: This should be `objectivefunctionvalues`

File: src/estimation/diagnostics.jl [hunk 2]
```diff
@@ -481,3 +481,106 @@ function DataFrames.DataFrame(i::FittedPumasModelInspection; include_covariates=
...
+
+Returns the observed function values during optimization.
```
Comment (vjd) [inline src/estimation/diagnostics.jl]: objective function

File: src/estimation/diagnostics.jl [hunk 3]
```diff
@@ -481,3 +481,106 @@ function DataFrames.DataFrame(i::FittedPumasModelInspection; include_covariates=
...
+Must return a `Vector{Number}`.
+"""
+_observedvalues(f::FittedPumasModel) = getproperty.(f.optim.trace, :value)
```
Comment (vjd) [inline src/estimation/diagnostics.jl]: _objectivefunctionvalues

File: src/estimation/diagnostics.jl [hunk 4]
```diff
@@ -481,3 +481,106 @@ function DataFrames.DataFrame(i::FittedPumasModelInspection; include_covariates=
...
+
+    data  = _convergencedata(arg) # again, a wrapper function.
+
+    layout --> good_layout(length(names) + 1)  # good_layout simply overrides the Plots layouter for n <= 4
+
+    legend := :none # clutters up the plot too much
+
+    # also plot the observed function value from Optim
+    @series begin
+        title := "Observed value"
```
Comment (vjd) [inline src/estimation/diagnostics.jl]: Objective Function Value

Comment (asinghvi17) [reply src/estimation/diagnostics.jl]: Will go through all of these and change, thanks!

**Comment by asinghvi17** [comment]
Not sure whether that would be the right approach if we were to make each folder a submodule?

**Pull request merged**